proposals/public-suffix.md

```diff
@@ -0,0 +1,555 @@
+|------------------------|-------------|------------|
+| Allow Private Suffixes | Yes         | including all suffixes in PSL means more information about third-party boundaries |
+| Allow Unknown Suffixes | Yes         | provides better performance |
+| Preserve IDN Unicode   | No          | provides better performance |
+
+### Known Consumers
+
+Mozilla intends to make use of this API in its [multi-account-containers](https://addons.mozilla.org/en-US/firefox/addon/multi-account-containers/)
+extension to allow users to create containers that isolate organizational base domains:
+[PR #2352](https://github.com/mozilla/multi-account-containers/pull/2352).
+Currently, users must manually curate their containers to capture all anticipated subdomains
+of an organization's base domain, and this manual approach is cumbersome and error-prone.
+
+In addition to this, an analysis has been done of the relevant use cases of a selection of
+other extensions that currently roll their own solutions in order to parse the PSL, and the
+use cases are set out in the [Use Cases](#use-cases) section. Although the extension
+authors have not been consulted directly, it is assumed that they would stand to benefit
+from this API due to the reasons stated in the [Objective](#objective) section.
+
+| Extension | Relevant use cases | Description |
+|-----------|--------------------|-------------|
+| [multi-account-containers](https://addons.mozilla.org/en-US/firefox/addon/multi-account-containers/) | #1, #2 | isolate requests into containers by organizational base domain, view all domains grouped by organization |
+| [uBlock Origin](https://addons.mozilla.org/en-US/firefox/addon/ublock-origin/) | #1, #2, #3 | create organization/3rd-party request-blocking rules, view request blocking status by organization, create blocking rules for specifically 3rd-party requests |
+| [adblock-plus](https://addons.mozilla.org/en-US/firefox/addon/adblock-plus/) | #1, #3 | create organization/3rd-party request-blocking rules |
+| [duckduckgo-for-firefox](https://addons.mozilla.org/en-US/firefox/addon/duckduckgo-for-firefox/) | #3 | exclude requests from tracking detection unless they are 3rd-party |
+| [violentmonkey](https://addons.mozilla.org/en-US/firefox/addon/violentmonkey/) | #1 | run scripts automatically for requests having a specific organization name plus any suffix |
+
+## Specification
+
+### Schema
+
+A new API `publicSuffix` is added as follows:
+
+```ts
+namespace publicSuffix {
+  //
+  // Object containing both registrable domains (i.e. ICANN/unknown, Private)
+  // for a domain.
+  //
+  interface RegistrableDomains {
+    // The original domain whose registrable domain we want
+    domain: string,
+    // The ICANN-suffixed or unknown-suffixed registrable domain.
+    // Null if an error occurred.
+    base?: string,
+    // The Private-suffixed registrable domain.
+    // Null if an error occurred, or if the domain has no matching Private suffix.
+    private?: string,
```

Comment:
Calling the API twice with `excludePrivateSuffixes` set to true and false has the following potential issues:

**Duplication of work**
* Must parse/canonicalize the same input domain parameter on each API call.
* The PSL lookup involves removing labels in turn from the input domain and testing the remaining suffix until a match is found. Therefore on the second API call, the same unmatching candidate suffixes that have already been tested in the first API call will again be retested. The difference will be that the algorithm keeps going further in terms of removing labels with the `excludePrivateSuffixes=true` call.

**Duplication of returned arrays**
* If the registrable domains array obtained with `excludePrivateSuffixes=false` happens to only contain ICANN domains (because no matching private suffixes exist), then the second API call with `excludePrivateSuffixes=true` will return exactly the same registrable domains array again.

Note that my proposal does not require "two lookups". One possible implementation would simply continue removing labels after finding a private suffix until the ICANN suffix is found, i.e. not a full lookup, but a continuation of the current lookup. Even better, a more optimal implementation would have the corresponding ICANN suffix pre-calculated and stored with every private suffix, such that no further work would be needed to determine the ICANN suffix upon matching a private suffix.

I am happy to remove `domain` in the result, because I agree this can be inferred by the array position.